### DISCUSSION TITLE
Date and Time Data Type

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#530] [Maintainer discussions]
The specs related to issue #240 are ready for review. Please take a look at the attached document and follow the [review guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines).

## Document Link
https://wiki.mathesar.org/en/design/specs/date-time-data-type

@mathemancer @kgodey @pavish, your feedback is required.
The deadline for feedback is set to a week from now, the first review round will be completed by Thursday 12th. Please plan your review so that at least one round of changes can be made and reviewed if needed.

**Comment by kgodey** [comment]
@ghislaineguerin Good start! Feedback:

(1) There's no option to show just time fields (without date).

(2) Related to the above, I'm not sure that "Show time" belongs in Advanced Options, since it's pretty fundamental to the type. If you think it would be better, we could have separate types for "Date", "Time", and "Date & Time".

(3) Timezones should be enabled by default, you have them as off by default.

(4) I don't think we need to allow the user to input date ranges while grouping. We can just say "Group by year" or "Group by quarter". If they want to only group specific dates, they can filter first.

(5) There is no option to support non-UTC timezones in the DB, so we don't need to support the option of setting a timezone.

(6) The prototype still shows a database type when advanced options is not expanded.

(7) There doesn't appear to be a prototype for the grouping options, the link just shows a Figma file that's open with only one screen.

(8) There's no design for a time picker.

(9) We need to define the date format options in more detail.

---

Also, general feedback on the review request format, you could shorten the last paragraph to:

> [USERNAMES], your feedback is required by [DATE]. 

It's a nitpick, but it's faster to read and there's no planning needed for future review rounds.

**Reply by ghislaineguerin** [reply]
Thanks, Kriti
1. I've added display options for the date/time type that allows users to choose between 'date,' 'date and time,' and 'time.'
2. I agree. It is now visible when the type is set to date/time
3. Agree, I have updated the prototype
4. I have updated the group options to reflect this change
5. Ok, changed
6. I have moved it under advanced
7. We no longer need that document. I will update the specs
8. There was one, you have to click on the time when editing a date/time field
9. Done. I have added detail for both date and time formats.

**Reply by kgodey** [reply]
@ghislaineguerin The update addresses most of my feedback, thanks! Some more thoughts:

(1) What you have as Date & Time display options actually change the underlying database type (similar to the "number type" options). I think we should reserve the term "display options" for things that only affect the UI and not the data stored in the database.

In general, words like "display" and "format" should be reserved for UI options, words like "type" and "values" should be reserved for changing things at the database level.

I think the solution you have works fine otherwise. I do think we should take a look at it in the context of all the other data types, since the "set number type" is similar in function to it (but also different, because number type is not essential like this one is).

(2) I like the short/long/medium dates and times, but I 
think we'll need more date formats to account for US vs. European vs. ISO dates. As an example, here's the format picker from Airtable:

![Screen Shot 2021-08-13 at 5 50 23 AM](https://user-images.githubusercontent.com/287034/129339486-abc7c730-4728-4063-9d8a-4cbec6b453f0.png)

We might also want to do 12 vs. 24 hour time in addition to what we have, here's also the example from Airtable

![Screen Shot 2021-08-13 at 5 50 37 AM](https://user-images.githubusercontent.com/287034/129339585-a6d3e466-13ef-487e-8118-ea762fd80cd1.png)

(3) Also while looking at Airtable, they have an option to show local times or fix the times to GMT, that might be something to consider:

![Screen Shot 2021-08-13 at 5 50 46 AM](https://user-images.githubusercontent.com/287034/129339725-00e6319d-2aa6-4a7e-a344-dbdc41dfba42.png)

(4) I still cannot find the timepicker.

(5) I think we need all the grouping options documented somewhere, even if just in text on the spec.

Thanks!

**Reply by pavish** [reply]
I agree with all of @kgodey's points here.

When timezones are enabled, we need to decide on how the data is shown on the UI. This is the same as point (3) above. This is not applicable when timezones are not enabled.

**Reply by ghislaineguerin** [reply]
@kgodey 
1. I agree, I also see this called 'Sub-type' in some contexts, something to consider when we look at all dropdowns together.
2.a. This might not work with the short/medium/long options, I will update according to your suggestion. 
2.b. I already had 12 and 24 hours but maybe it wasn't clear, the tt part of the time determines that. I will update.
3. Good idea, we'd need a way to indicate this in the dropdown outside of the type options.
4. Should be there now, I think the prototype was not pointing to that input field.
5. Agreed. I will work on that.

**Reply by ghislaineguerin** [reply]
@kgodey Going back to the date/time options and why I sometimes choose to show how the actual date is constructed, so for example, M/d/yy vs. 8/13/2021 in your Airtable example. One of our design goals is to avoid over-abstraction as we want users to eventually manage their database or look at their data and make sense of it. When I look for options in presenting things, I try to find the less abstract but still usable way to present them, rather than going completely 'user-friendly'. I do get your point that we need additional options that are regularly used. I just wanted to clarify my intent when considering the options.

**Reply by kgodey** [reply]
@ghislaineguerin I'm fine with showing the option as `M/d/yy`, the Airtable example was just to show the other date formats available. I don't think we should copy the way they present it.

**Comment by mathemancer** [comment]
Here's a discussion with a little context about Time zones:

https://github.com/centerofci/mathesar/discussions/119

I don't quite understand what the 'support time zones' toggle is intended to achieve.  Is it one (or all) of the things in the discussion, or something else?

For me, it's not clear what the manual entry will look like.  Will they be typing in anything, or will we expecting to "train" them (through UI elements) to input times / dates in a particular format?

**Reply by kgodey** [reply]
@mathemancer "Support time zones" is intended to enable the "with time zone" option on `timestamp` and `time` Postgres types: https://www.postgresql.org/docs/current/datatype-datetime.html

**Comment by ghislaineguerin** [comment]
@mathemancer @kgodey @pavish 
This is ready for an additional review.

**Comment by ghislaineguerin** [comment]
@kgodey @pavish 
The document has been updated to include the global date-time options. This update addresses the following:
- Date and Time format options
- Clear separation between database and display options

**Reply by kgodey** [reply]
Looks good, thanks! I assume showing what timezone to use for the column (local time or UTC) will be handled separately?

**Comment by ghislaineguerin** [comment]
@kgodey
After our discussion on Friday, @pavish had some concerns over inconsistent controls for date/time and duration. He proposed that we use autocomplete functionality to add time units and allow users to customize the resulting format.
 
![image](https://user-images.githubusercontent.com/845767/130427408-f8b48c83-c561-4793-9cd5-a0b90b18efcc.png)

Where activating the dropdown would provide the user with a time units selector, that once added, can be further modified by specifying its format.

![image](https://user-images.githubusercontent.com/845767/130427589-12973228-d9b6-4cad-bd0d-e707cef4b6dc.png)

This is similar to what other spreadsheet apps have. However, I will leave it up to @pavish to decide if the functionality as it is presented doesn't add more complexity to the implementation.

**Reply by kgodey** [reply]
@ghislaineguerin This seems like a powerful option for advanced users but it loses the simplicity of the previous controls for less technical users (and I think it might overwhelm them with the number of controls). I would rather go with a simple implementation at first.

Could you or @pavish document in detail what the original concerns were? It's difficult to come up with good feedback for a design change if I don't understand what prompted it.

**Reply by pavish** [reply]
@ghislaineguerin I meant autocomplete, as in a single text box, with no additional controls. We can show a list of suggested options and users can select one of them/type in their own option. There are no units as such, it could be any text that they input.

I agree with @kgodey that the new option makes it more complicated. Here are my original concerns:
* When it comes to date/time formats, people are used to entering the formats directly. The previous design worked solely on checkboxes, which is not a common design pattern and might confuse users.
* Customising the format would not be possible when we depend only on checkboxes.
* The units are not necessarily important. Users may want them, or may not. Having them type their own format/choose from a list would make it way easier than forcing them to stick to our patterns.

**Reply by kgodey** [reply]
@pavish Your feedback seems to be for the duration control, not the date/time controls.

**Reply by pavish** [reply]
@kgodey My feedback applies to both date/time and duration. I do think only a single text box, with suggestions would suffice for both. This is the most common interface used for specifying date/time formats (From operating systems to spreadsheet applications), and I think anything else would only make it more complicated or reduce functionality.

**Reply by kgodey** [reply]
@pavish Your concerns focused on checkboxes, I don't think the date/time specs had checkboxes previously. If you have additional concerns with the previous version of date/time specs, please specify those too.

Could you provide an example of the single text box interface? I looked at a few different applications and couldn't find what you were referring to. Here's what I found:

### Airtable
![Screen Shot 2021-08-23 at 9 00 58 AM](https://user-images.githubusercontent.com/287034/130451601-3b810304-a2fe-485d-b552-fa0b02840131.png)
![Screen Shot 2021-08-23 at 9 00 51 AM](https://user-images.githubusercontent.com/287034/130451602-dcca14b7-c39b-4789-bb2f-c05b8af45ec7.png)

### macOS
![Screen Shot 2021-08-23 at 8 54 16 AM](https://user-images.githubusercontent.com/287034/130451272-806e4373-d43f-417b-a606-febfb1701e28.png)

### Numbers
![Screen Shot 2021-08-23 at 8 57 55 AM](https://user-images.githubusercontent.com/287034/130451316-abd44099-84e4-4d79-a526-ff7bd763d29a.png)
![Screen Shot 2021-08-23 at 8 57 41 AM](https://user-images.githubusercontent.com/287034/130451317-7f2f1af1-fb68-4bf7-a358-844d2d453bfd.png)

### Google Sheets
![Screen Shot 2021-08-23 at 8 40 57 AM](https://user-images.githubusercontent.com/287034/130451385-ee7941a7-b03d-487b-848e-6fb7bc4a1e60.png)

**Reply by pavish** [reply]
@kgodey 
* I assumed that we needed to provide an option to customize the date/time format (other than the available presets), for the MVP. If we do not require it for the MVP, the previous design is good for me. My whole argument was based on the idea that we had to provide the option.
* The following images illustrate the thought I had in mind.
  * Taken from Windows/Excel, Zoho sheets.
    ![custom-date-format-excel](https://user-images.githubusercontent.com/11850603/130976954-37bdddf7-95ff-4929-b9e1-ef6c3cf67f5a.png)
    ![3](https://user-images.githubusercontent.com/11850603/130976975-90c25a5a-75b0-4fbf-9550-c2b32d3caf7b.png)

### A sample mockup of what I had in mind:
  ![Group 20](https://user-images.githubusercontent.com/11850603/130977033-ef40ce56-d0a7-4157-ae73-e3e9e3a3a160.png)

* The user would have the option to choose from available presets, or enter their own format. When the preset is clicked, it updates the text field. There is a preview option which shows the current date or time in the entered format, or if the format is invalid, the space will be used to show an error.
* User can specify both date and time together (For the date+time db type). We would not need separate controls for date and time.
* This will allow the user to specify any combination, separators, or even additional text, along with the date/time.
* However, this change would also require us to rethink the UX for editing the date/time cell on the table. 

### Duration and Date/time
* During the call with @ghislaineguerin , we discussed both duration and date/time formats together, and my initial idea was that it would be better if we had a similar interface for both. However, after quite a bit of thought, I think duration needs to be handled differently from date/time. We can discuss duration specific thoughts in https://github.com/centerofci/mathesar/discussions/536.

**Reply by kgodey** [reply]
@pavish Thanks for explaining! For the MVP, I think it would be simplest for us to provide a set of presets and not allow the user to customize beyond that. 

In the future, if we do want to allow the user to customize the format, I like the option you've suggested and we can evaluate it then.

**Comment by ghislaineguerin** [comment]
@kgodey @pavish I have reverted the date/time format controls to use the original dropdown design
